python/lsst/dax/obscore/obscore_exporter.py

```diff
@@ -307,6 +326,129 @@ def to_csv(self, output: str) -> None:
+        votable.resources.append(resource)
+
+        fields = []
+        for arrow_field in self.schema:
+            if arrow_field.name in obscore_columns:
+                ffield = obscore_columns[arrow_field.name]
+                votable_datatype = FelisType.felis_type(ffield.datatype.value).votable_name
+                field = astropy.io.votable.tree.Field(
+                    votable,
+                    name=ffield.name,
+                    datatype=votable_datatype,
+                    arraysize=ffield.votable_arraysize,
+                    unit=ffield.ivoa_unit,
+                    ucd=ffield.ivoa_ucd,
+                )
+                fields.append(field)
+            elif arrow_field.name == "em_filter_name":
+                # Non-standard but part of internal standard schema.
+                field = astropy.io.votable.tree.Field(
+                    votable,
+                    name="em_filter_name",
+                    datatype="char",
+                    arraysize="*",
+                )
+                fields.append(field)
+            else:
+                # This must be a non-standard field. Attempt to add it.
+                kwargs = {}
+                datatype = ""
+                if (
+                    arrow_field.type.equals(pyarrow.int64())
+                    or arrow_field.type.equals(pyarrow.int32())
+                    or arrow_field.type.equals(pyarrow.int16())
+                ):
+                    datatype = "int"
+                elif arrow_field.type.equals(pyarrow.string()):
+                    datatype = "char"
+                    kwargs["arraysize"] = "*"
+                elif arrow_field.type.equals(pyarrow.float64()):
+                    datatype = "double"
+                else:
+                    raise RuntimeError(f"Could not handle unrecognized column {arrow_field.name}")
+                field = astropy.io.votable.tree.Field(
+                    votable,
+                    name=arrow_field.name,
+                    datatype=datatype,
+                    **kwargs,
+                )
+                fields.append(field)
+
+        table0 = astropy.io.votable.tree.TableElement(votable)
+        resource.tables.append(table0)
+        table0.fields.extend(fields)
+
+        def is_none(v: Any) -> bool:
+            return v is None
+
+        chunks = []
+        n_rows = 0
+        for record_batch in self._make_record_batches(self.config.batch_size, limit=limit):
+            pydict = record_batch.to_pydict()
+            columns = []
+            for label, column in pydict.items():
+                # Need to mask out any None values.
+                mask = [is_none(v) for v in column]
+                mc = astropy.table.MaskedColumn(column, name=label, mask=mask)
+                columns.append(mc)
+            chunk = astropy.table.Table(columns)
+            n_rows += len(chunk)
+            array = ma.array(np.asarray(chunk), mask=np.asarray(chunk.mask))
```

Comment:
@erykoff do you have any advice on going from RecordBatch to numpy masked array that can be stacked? The whole "create MaskedColumn, create Table from MaskedColumn, then convert back to numpy masked array" approach is naive and causes an obvious hotspot. I can't help but feel that there is a better way to go from the parquet RecordBatch to the masked array without going through the astropy table so your thoughts would be appreciated.

Reply:
I've never used a `RecordBatch` before, so can't say much off the top of my head.  I do see that they can be transformed into a dataframe (but you probably don't want pandas here).  I wonder if `lsst.daf.butler.formatters.parquet.arrow_to_numpy()` would work (or with some modification).

Reply:
It looks like the fundamental problem is that `RecordBatch` doesn't seem to know about nulls so `nullcount` doesn't work. I will see if I can use hints from `arrow_to_numpy` to end up in the same place without going via the astropy table.

Reply:
I've made some tweaks to `arrow_to_numpy` on lsst/daf_butler#1087 and this cleans up all this code (converting to arrow Table first) and halves the run time of the query tests. Now all the overhead is doing actual queries.